### PR TITLE
fix(trtllm): apply full stop conditions during PREFILL, not just max_tokens

### PR DESCRIPTION
**Pull request by peilii**
## Summary

- `PREFILL` forces `max_tokens=1` (context-only) but was gating the rest of stop-condition handling on `is_generation_stage(...)`, which `PREFILL` never satisfies. As a result `ignore_eos` (and the rest of `stop_conditions`) was silently dropped during prefill.
- When a caller requested `ignore_eos=true` and the prompt's first sampled token happened to be EOS, TRT-LLM treated it as a stop condition on that single forced token and terminated the request right there instead of letting the KV handoff to decode proceed. The request completed with HTTP 200 and `output_tokens=1`, but no visible content was ever produced.
- This regressed in `d3e815a9c` (#11671), which added the `is_generation_stage` gate around `apply_stop_conditions_to_sampling_params(...)`.
- Found via internal testing.

## Fix

Apply `apply_stop_conditions_to_sampling_params(...)` for `PREFILL` as well as generation-stage workers, matching native TRT-LLM `context_only` behavior: only `max_tokens` is overridden, the rest of the request's stop parameters (`min_tokens`, `ignore_eos`, `stop_token_ids`) pass through unchanged.

## Validation

- Updated `test_prefill_skips_generation_stop_conditions` (renamed to `test_prefill_applies_stop_conditions_matching_native_context_only`) to assert `min_tokens`, `ignore_eos`, and `stop_token_ids` are now applied during prefill, matching native `context_only` semantics.
- Added `test_prefill_does_not_force_ignore_eos_when_not_requested` covering the negative case: `ignore_eos` stays `False` when not requested, while `min_tokens`/hidden stop tokens are still preserved.
- `isort`, `black`, `flake8`, `codespell`, `ruff`, and the repo's `pytest-marker-report` pre-commit hook all pass on the changed files.
- Could not run the full pytest suite locally (no GPU/`torch` in this environment); relying on CI for execution.

---

<a href="https://nvidia.devinenterprise.com/review/ai-dynamo/dynamo/pull/13442" target="_blank">
  <picture>
    <source media="(prefers-color-scheme: dark)" srcset="https://static.devin.ai/assets/gh-open-in-devin-review-dark.svg?v=1">
    Open in Devin Review
  </picture>
</a>

## Summary by CodeRabbit

* **Bug Fixes**
  * Stop conditions are now consistently applied during both prefill and generation stages.
  * Prefill requests continue to limit output to one token while preserving other settings, including minimum tokens, EOS handling, and stop-token behavior.
  * Explicitly setting EOS handling to false is now preserved correctly.

### CODE REVIEW

**Review by richardhuo-nv — APPROVED** [review]

**Review by nv-kmcgill53 — COMMENTED** [review]
Can you double check that this doesn't break in aggregated mode?

File: components/src/dynamo/trtllm/request_handlers/handler_base.py [hunk 1]
```diff
@@ -1130,7 +1130,19 @@ async def _generate_locally_impl(
...
+        # prompt whose first token happened to be EOS terminated the request
+        # right there instead of letting the KV handoff to decode proceed --
+        # the request "succeeded" with zero visible output instead of
+        # continuing generation normally.
+        if (
+            is_generation_stage(CommonDisaggregationMode[self.disaggregation_mode.name])
+            or self.disaggregation_mode == DisaggregationMode.PREFILL
+        ):
             apply_stop_conditions_to_sampling_params(
                 sampling_params, request["stop_conditions"]
```
[the remark this answers is not in the document]
Comment (peilii) [reply components/src/dynamo/trtllm/request_handlers/handler_base.py:1142]: Checked against the pinned TensorRT-LLM version (v1.3.0rc23) and the exact backend our recipes use (PYTORCH). `min_tokens`/`min_length` is implemented purely as a logit-masking loop in `_torch/pyexecutor/sampler/sampler.py::_apply_min_length_penalty`, bounded by the actual number of decode steps being executed:

```python
for step in range(num_steps[index]):
    if (r.get_num_tokens(beam_idx) - r.py_orig_prompt_len) + step < r.py_min_length[0]:
        rows.append(...); cols.append(end_id)  # mask EOS logit
    else:
        break
```

For PREFILL, `num_steps=1` (forced by `max_tokens=1`), so this just masks EOS on that single step if `min_length > 1`  -- there's no comparison against `max_new_tokens`/`max_tokens` anywhere in this function, and none in `SamplingParams._validate()` either. So `min_tokens=8` with `max_tokens=1` doesn't raise or get rejected -- it's a no-op beyond suppressing EOS on the one step that runs, which is consistent with (not in conflict with) what `ignore_eos` already does.

**Comment by peilii** [comment]
@nv-kmcgill53 Confirmed this doesn't affect `AGGREGATED` mode, for two independent reasons:

1. The `max_tokens=1` override only fires for `disaggregation_mode == PREFILL` (a separate, untouched block) -- `AGGREGATED` requests never hit it, so `max_tokens` stays whatever the caller sent, same as before this PR.
2. The stop-conditions gate is now `is_generation_stage(mode) or mode == PREFILL`. `AGGREGATED` (and `DECODE`) were already in `_GENERATION_STAGES`, so `is_generation_stage(AGGREGATED)` was already `True` before this change -- the added `or mode == PREFILL` clause is redundant for `AGGREGATED` and never changes the outcome. It takes the exact same branch, calling `apply_stop_conditions_to_sampling_params` exactly as it did before.

The full diff to `handler_base.py` is just this one condition change -- nothing else touched. There's also existing `AGGREGATED`-mode coverage in `test_trtllm_handler_base.py` (e.g. `test_health_check_gets_priority_1`) exercising the same `generate_locally` path, which stays unaffected since the change is a no-op for that mode.

File: components/src/dynamo/trtllm/request_handlers/handler_base.py [hunk 2]
```diff
@@ -1130,7 +1130,19 @@ async def _generate_locally_impl(
             if default_max_tokens is not None:
                 sampling_params.max_tokens = default_max_tokens
 
-        if is_generation_stage(CommonDisaggregationMode[self.disaggregation_mode.name]):
+        # PREFILL forces max_tokens=1 above but must still apply the full stop
```
Comment (tanmayv25) [inline components/src/dynamo/trtllm/request_handlers/handler_base.py]: Too verbose. Can you simplify?

Comment (peilii) [reply components/src/dynamo/trtllm/request_handlers/handler_base.py]: Trimmed to 3 lines in 03f9bb4 -- kept just the why (native context_only parity, avoid a false EOS stop skipping the KV handoff to decode).
